seqr/views/apis/anvil_workspace_api_tests.py

```diff
@@ -412,40 +452,191 @@ def test_create_project_from_workspace(self, mock_tempfile, mock_mv_file, mock_s
         mock_mv_file.assert_called_with(
             TEMP_PATH, 'gs://seqr-datasets/v02/GRCh38/AnVIL_WES/{guid}/base/{guid}_ids.txt'.format(guid=project.guid),
             user=self.manager_user
         )
 
+        self._assert_dag_trigger_requests(responses, UPDATED_ANVIL_VARIABLES, UPDATE_DAG_TASKS_RESP, 3, project.guid)
+
+        self._assert_sent_slack_message(project, mock_slack, mock_send_email, 'GRCh38')
+
+        # Test project exist
+        url = reverse(create_project_from_workspace, args=[TEST_WORKSPACE_NAMESPACE, TEST_NO_PROJECT_WORKSPACE_NAME])
+        response = self.client.post(url)
+        self.assertEqual(response.status_code, 400)
+        self.assertEqual(response.reason_phrase, 'Project "{name}" for workspace "{namespace}/{name}" exists.'
+                         .format(namespace=TEST_WORKSPACE_NAMESPACE, name=TEST_NO_PROJECT_WORKSPACE_NAME))
+
+        # Test saving ID file exception
+        responses.calls.reset()
+        url = reverse(create_project_from_workspace, args=[TEST_WORKSPACE_NAMESPACE, TEST_NO_PROJECT_WORKSPACE_NAME2])
+        mock_mv_file.side_effect = Exception('Something wrong while moving the ID file.')
+        # Test triggering dag exception
+        responses.replace(responses.GET,
+                      '{}/api/v1/dags/seqr_vcf_to_es_AnVIL_WES_v0.0.1/dagRuns'.format(MOCK_AIRFLOW_URL),
+                      json=DAG_RUNS_RUNNING)
+
+        response = self.client.post(url, content_type='application/json', data=json.dumps(REQUEST_BODY))
+        self.assertEqual(response.status_code, 200)
+        project2 = Project.objects.get(workspace_namespace=TEST_WORKSPACE_NAMESPACE, workspace_name=TEST_NO_PROJECT_WORKSPACE_NAME2)
+
+        self._assert_loading_req_exceptions(
+            mock_api_logger, mock_airtable_logger, mock_slack, mock_send_email, project2, "GRCh38",
+            samples=['HG00735', 'NA19675', 'NA19678']
+        )
+
+        # Test logged in locally
+        remove_token(self.manager_user)  # The user will look like having logged in locally after the access token is removed
+        response = self.client.post(url)
+        self.assertEqual(response.status_code, 302)
+        self.assertEqual(response.url, '/login/google-oauth2?next=/api/create_project_from_workspace/my-seqr-billing/anvil-no-project-workspace2/submit')
+
+    @mock.patch('seqr.views.apis.anvil_workspace_api.ANVIL_LOADING_DELAY_EMAIL', None)
+    @responses.activate
+    def test_add_workspace_data(self, mock_send_email, mock_datetime, mock_api_logger, mock_tempfile, mock_mv_file,
+                                mock_slack, mock_load_file, mock_airtable_logger, mock_utils_logger):
+        self._setup_dag_trigger(responses, ADD_DATA_UPDATE_DAG_TASKS_RESP)
+
+        # Test requesting to load data from a workspace without an existing project
+        url = reverse(add_workspace_data, args=['no_PROJECT1_GUID'])
+        self.client.force_login(self.collaborator_user)
```

Comment:
Why aren't you `self.check_collaborator_login` in this function? We always start our tests with a check login call on a real URL, and then do the error tests

Reply:
I need to login first to test the behavior for the non-existing project. But I found that I should do the `insufficient Anvil permissions` test first. Then no explicit `force_login` is required.